runtime/src/bank.rs

```diff
@@ -4003,7 +4003,21 @@ impl Bank {
     }
 
     pub fn get_account_modified_slot(&self, pubkey: &Pubkey) -> Option<(AccountSharedData, Slot)> {
-        self.rc.accounts.load_slow(&self.ancestors, pubkey)
+        self.load_slow(&self.ancestors, pubkey)
+    }
+
+    fn load_slow(
+        &self,
+        ancestors: &Ancestors,
+        pubkey: &Pubkey,
+    ) -> Option<(AccountSharedData, Slot)> {
+        // get_account (= this fn caller) can be called for various on-consensus
+        // code like inflation, feature activation and native program loading
+        // before freezing and it's generally impossible to make static/lexical guarantee of RPC or
+        // on-consensus.. so just pass !freeze_started for is_root_fixed as a close approximation
```

Comment:
this is another potential bug fix.

Reply:
hmmm I'm worried about hiding this behind `bank.freeze()`. There's no guarantee in the future that we won't need to load something from a frozen bank where the `load` is not allowed to fail (some services just keep a bank around as an accessor to the AccountsDb). 

I'm thinking it should be up to the caller, like RPC, to decide if a failure is acceptable.

Reply:
So, it took some back-and-forth by myself. Here's my revamped solution with another beefy commit: e4590ebae67ba4884c69c84c50e84c3b9344c3be

Let's reframe the situation a bit.

Namely, we can elide the externally observable race conditions altogether, _if we re-interpret the is_root_fixed flag to merely be an optimization hint and make the RPC loading route to newly-added slow path inside the index lock._

First of all, plumbing is_root_fixed up to RPC (as suggested by you: https://github.com/solana-labs/solana/pull/15454#discussion_r607473929) is very tedious(77010dbbe2f2f2cd73b228bf5a0e86f90632eac4). Also, the outcome is that rpc remain to be exposed to the race condition (which is very undesirable, come to think of it), even including very fundamental methods like `getAccount` with the `confirmed/finalized` commitment while holding the bank correctly. That's bad. we're just externalizing our bad race bug just to others shoulders. I think rpc should be treated as a first-citizen. Also, I don't want to handle supporting tickets for these insane flaky rpc behavior and recommend partners to retry even getAccount as it could _be silently faulty_ (returning empty account at the definitely existing address), which sounds pretty lame. ;) Maybe unrelated to this, As a illustration, I want to avoid these hard-to-debug rpc bug reports: https://github.com/solana-labs/solana/issues/15681, people would eventually encounter in the future if we had this data race rather permanent thing.

Finally, about some supportive arguments. Firstly, I think it's acceptable from RPC consumer's perspective for rpc to be slightly slow and to affect on-consensus processing a bit, rather than the hard-to-debug data-race bugs. (so this is revised version of my previous failed suggestion: https://github.com/solana-labs/solana/pull/15454#discussion_r590759082).

Also, I believe newer implementation for rpc load doesn't cause problems:
- I made relevant `match` arms be `unreachable!()`, removing the need of `num_acceptable_failed_iterations` and making `retry_to_get_account_accessor` be used only for `LoadSafety::FixedMaxRoot` (this makes things simpler)
- With the assumption no purge race (e02974b27a7afcafdfa34b54afcd69c072dc17c5 ecdf6ae1bc981a1884135482394ee818097f8ae0) with RPC, we can prove that the new rpc load doesn't fail anymore, as long as we hold the corresponding lock in read_index_for_accessor_or_load_slow.
- replaying/banking works as fast as before if there is no rpc access. :)

@carllin How do you like this?

Reply:
hmmmm, I'm still worried about the perf implications of loading these large accounts under heavier RPC load. I'm especially hesitant to incur these costs, since these failures should be very rare (race means you would have to consistently encounter a new root between the accounts index fetch and the accounts load, especially since roots should not be being set more than once every few hundred ms). 

One thing we could do is get rid of the failure case altogether, and just keep retrying with some very absurdly high limit. I would be very surprised if this ever ended up in a loop that lasted more than a few iterations. We can keep the logging around to observe if this ever does happen.

Reply:
I see! thanks for quick suggestion. I'll adjust this pr yet again. :)

Reply:
@carllin hi, here's my another patch, believing it's done accordingly as suggested by you: https://github.com/solana-labs/solana/pull/15454/commits/b8ced7f7334a4e2f2558affd2b461c456f445a6b

Could you review this? thanks for in-depth review!

Reply:
This patch restores the previous behavior and I checked all of new tests fails with this:

```patch
$ git diff carllin/FixAccountsDb
diff --git a/runtime/src/accounts_db.rs b/runtime/src/accounts_db.rs
index 4fa24ede44..3799f73301 100644
--- a/runtime/src/accounts_db.rs
+++ b/runtime/src/accounts_db.rs
@@ -2499,6 +2499,8 @@ impl AccountsDb {
         let mut num_acceptable_failed_iterations = 0;
         loop {
             let account_accessor = self.get_account_accessor(slot, pubkey, store_id, offset);
+            return Some((account_accessor, slot));
+
             match account_accessor {
                 LoadedAccountAccessor::Cached(Some(_)) | LoadedAccountAccessor::Stored(Some(_)) => {
                     // Great! There was no race, just return :) This is the most usual situation
```

```
failures:
    accounts_db::tests::test_load_account_and_cache_flush_race_with_retry
    accounts_db::tests::test_load_account_and_cache_flush_race_without_retry
    accounts_db::tests::test_load_account_and_purge_race_with_retry
    accounts_db::tests::test_load_account_and_purge_race_without_retry
    accounts_db::tests::test_load_account_and_shrink_race_with_retry
    accounts_db::tests::test_load_account_and_shrink_race_without_retry

test result: FAILED. 0 passed; 6 failed; 0 ignored; 0 measured; 489 filtered out; finished in 30.10s

error: test failed, to rerun pass '-p solana-runtime --lib'

```

Reply:
Also, I'll run a test validator against mainnet-beta for a day or two before merging this to make sure everything works.

Reply:
@ryoqun looks good! I had one last remaining concern here: https://github.com/solana-labs/solana/pull/15454#discussion_r612045902.